File: 11/stones copy.py
IN = "125 17"
try:
    import json
    s, stones = json.loads(open("stones").read())
    c = 75
except:
    IN = "5910927 0 1 47 261223 94788 545 7771"
    s = 0
    c = 75
    stones = [int(x) for x in IN.split()]


def do(stones):
    i = 0
    while i < len(stones):
        stone = stones[i]
    # for i, stone in range(len(stones)):
        # print(i, stone)
        if stone == 0:
            stones[i] = 1
            i += 1
        elif len(str(stone)) % 2 == 0:
            del stones[i]
            s = str(stone)
            mid = int(len(s)/2)
            left, right = s[:mid], s[mid:]
            stones.insert(i, int(right))
            stones.insert(i, int(left))
            i += 2
        else:
            stones[i] = stone * 2024
            i += 1

    return stones

def split(a, n):
    k, m = divmod(len(a), n)
    return (a[i*k+min(i, m):(i+1)*k+min(i+1, m)] for i in range(n))

def flatten(xss):
    return [x for xs in xss for x in xs]

if __name__ == "__main__":
    try:
        import multiprocessing

        with multiprocessing.Pool(20) as p:
            for _ in range(s, c):
                print(_)
                stones = flatten(p.map(do, split(stones, 10)))

            # print(stones)

            # print(stones)
        # if _ == 1:
        #     break
        # print(stones)
        # breakpoint()

            print(_, 'stones', len(stones))
        print('stones', len(stones))
    except KeyboardInterrupt:
        import json
        open("stones", "w").write(json.dumps([_, stones]))

File: 11/stones copy.go
package main

import (
	"fmt"
	"strconv"
)

// var cache [2024*2024]int
var cache [76]map[int]int

func blink(stones []int, blinks int) int {
	// fmt.Printf("%d %+v\n", 6-blinks, stones)
	if blinks <= 0 {
		return len(stones)
	}

	count := 0
	var stone int
	for _, stone = range(stones) {
		if cache[blinks][stone] != 0 {
			// fmt.Printf("cache %d %d %+v\n\n\n\n", blinks, stone, cache)
			count += cache[blinks][stone]
			continue
		}
		if stone == 0 {
			if blinks == 1 {
				count += 1
				continue
			}
			if blinks == 2 {
				count += 1
				continue
			}
			i := blink([]int{20,24}, blinks - 3)
			count += i
			continue
		}
		// if stone < 2024*2024 && cache[stone] != 0 {
		// 	count += cache[stone]
		// 	continue
		// }
		if s := strconv.Itoa(stone); len(s) % 2 == 0 {
			mid := int(len(s)/2)
			left, _ := strconv.Atoi(s[:mid])
			right, _ := strconv.Atoi(s[mid:])
			i := blink([]int{left, right}, blinks -1)
			count += i
			continue
		}
		i := blink([]int{stone * 2024}, blinks - 1)
		// fmt.Printf("%d %d\n", stone, i)
		cache[blinks][stone] = i
		count += i
	}
	return count
}

func main() {
	// cache = [2048*2048]int{}
	for i := range(76) {
		cache[i] = make(map[int]int)
	}

	var stones []int

	// stones = []int{125,17}
	// fmt.Printf("22\t\t%d\n", blink(stones, 6))

	// fmt.Printf("55312\t\t%d\n", blink(stones, 25))


	// stones = []int{5910927,0,1,47,261223,94788,545,7771}
	// fmt.Printf("193607\t\t%d\n", blink(stones, 25))


	fmt.Printf("193607\t\t%d\n", blink(stones, 75))
}
